pageserver/src/tenant/storage_layer/layer.rs

```diff
@@ -666,6 +666,8 @@ impl LayerInner {
                         return Err(DownloadError::DownloadRequired);
                     }
 
+                    tracing::info!(%reason, "downloading on-demand");
```

Comment:
```suggestion
                    tracing::info!(%reason, task_kind=?ctx.map(|x| x.task_kind()), "downloading on-demand");
```

Perhaps. But I kind of dislike this because then you need disclaimers around this that the task kind we log is selected by luck.

One thing to consider is the observability of concurrent on-demand wants of a layer which is now degraded from the earlier PR, not that it was often used, but did solve us one production investigation.

Reply:
> One thing to consider is the observability of concurrent on-demand wants of a layer which is now degraded from the earlier PR, not that it was often used, but did solve us one production investigation.

This could be solved by each incrementing a counter on LayerInner before they enter `self.inner.get_or_init`, and the one initializing could log the value before returning?

Everyone would reset their incremented counters manually.

Reply:
I'll implement this still, just in case it solves another investigation.

Reply:
Hmm I should add more API for this, to avoid another atomic counter while not necessary. Will do another PR.

Reply:
waiter count and logging: [d1c6070](https://github.com/neondatabase/neon/pull/5766/commits/d1c607093f28ef1d2b3a8c3c19ddb41ec528472c)